006-improving-configurability-of-kafka-listeners.md

```diff
@@ -0,0 +1,240 @@
+One of the examples is joining their Kubernetes network with their network outside of Kubernetes.
+In such case, a common problem is using or not using the cluster service DNS suffix (`.cluster.local` by default) which is in some cases needed and in some not.
+Having the ability to override the advertised hostnames on internal listeners similarly to how they can be treated on the external listener or being able to have ore internal listeners could help in these situations.
+
+## Proposed changes
+
+This proposal tries to make the configuration of the listeners more flexible.
+It should use array / list instead of an object with predefined listeners.
+It should also open more configuration options to all of them.
+
+_Note: The replication listener should be still kept non-configurable and hidden from users._
+
+The new configuration can look like this:
+
+```yaml
+    listeners:
+      - name: encrypted1
+        port: 9092
+        type: service
+        tls: true
+        authentication:
+          type: tls
+        overrides:
+          brokers:
+            - broker: 0
+                advertisedHost: my-cluster-brokers-kafka-0.myns.cluster.local
+            - broker: 1
+                advertisedHost: my-cluster-brokers-kafka-1.myns.cluster.local
+            - broker: 2
+                advertisedHost: my-cluster-brokers-kafka-2.myns.cluster.local
+      - name: encrypted2
+        port: 9093
+        type: service
+        tls: true
+        authentication:
+          type: acram-sha-512
+      - name: routes
+        port: 9094
+        type: route
+        tls: true
+        authentication:
+          type: tls
+      - name: nodeports
+        port: 9095
+        type: nodeport
+        tls: true
+        authentication:
+          type: tls
+```
+
+The listeners will be configurable as array.
+Each listener will have its own unique name and will specify the port and type as required values.
+The port can be set to anything apart from `9091` (internal replication listener) and `9404` (Prometheus).
+A new type service will be introduced for the internal listeners designed for apps running inside the same Kubernetes cluster.
+Additionally, all the types from the existing external listeners will be supported as well.
+
+together with this change, I suggest to change the default value of the listener `tls` flag from `true` to `false`.
+The current situation, when TLS is enabled by default without using the `tls` field and needed to explicitly use the `tls: false` to disable TLS seems unintuitive and confusing.
+
+### Backwards compatibility
+
+The old format can be easily converted into the new format without any information loss.
+The example YAML below corresponds to the first example from the _Current situation_ section:
+
+```yaml
+    listeners:
+      - name: plain
+        port: 9092
+        type: service
+        tls: false
+        authentication:
+          type: scram-sha-512
+        networkPolicyPeers:
```

Comment:
Why do the network policy peers and authentication get removed in the example below?

Reply:
The second example show how would 
```yaml
    listeners:
      plain: {}
      tls: {}
```
be converted. So the network policy peer arrays are empty nd not used.